eZ/Publish/Core/FieldType/Image/ImageStorage/Gateway/DoctrineStorage.php

```diff
@@ -311,4 +271,70 @@ public function extractFilesFromXml($xml)
 
         return null;
     }
+
+    /**
+     * Checks whether image reference for given version already exists.
+     *
+     * @param int $fieldId
+     * @param \eZ\Publish\SPI\Persistence\Content\VersionInfo $versionInfo
+     *
+     * @return bool
+     */
+    private function imageReferenceExistsForVersion(int $fieldId, VersionInfo $versionInfo): bool
+    {
+        $selectQuery = $this->connection->createQueryBuilder();
+        $selectQuery
+            ->select($this->connection->quoteIdentifier('id'))
+            ->from($this->connection->quoteIdentifier('ezimagefile'))
+            ->where(
+                $selectQuery->expr()->andX(
+                    $selectQuery->expr()->eq(
+                        $this->connection->quoteIdentifier('contentobject_attribute_id'),
+                        ':fieldId'
+                    ),
+                    $selectQuery->expr()->eq(
+                        $this->connection->quoteIdentifier('version'),
+                        ':versionNo'
+                    )
+                )
+            )
+            ->setParameter(':fieldId', $fieldId, PDO::PARAM_INT)
+            ->setParameter(':versionNo', $versionInfo->versionNo, PDO::PARAM_INT);
+
+        $statement = $selectQuery->execute();
+
+        return (int) $statement->fetchColumn() !== 0;
+    }
+
+    /**
+     * Updates existing image reference for given version.
+     *
+     * @param string $uri
+     * @param int $fieldId
+     * @param \eZ\Publish\SPI\Persistence\Content\VersionInfo $versionInfo
+     */
+    private function updateImageReferenceForVersion(string $uri, int $fieldId, VersionInfo $versionInfo): void
+    {
+        $updateQuery = $this->connection->createQueryBuilder();
+        $updateQuery
+            ->update($this->connection->quoteIdentifier('ezimagefile'))
+            ->set($this->connection->quoteIdentifier('filepath'), ':filepath')
+            ->where(
+                $updateQuery->expr()->andX(
+                    $updateQuery->expr()->eq(
+                        $this->connection->quoteIdentifier('contentobject_attribute_id'),
+                        ':fieldId'
+                    ),
+                    $updateQuery->expr()->eq(
+                        $this->connection->quoteIdentifier('version'),
+                        ':versionNo'
+                    )
+                )
+            )
+            ->setParameter(':filepath', $uri, PDO::PARAM_STR)
```

Comment:
For the entirely new `DoctrineStorage` methods please use `ParameterType::*` instead. 
For additions to the existing methods and `LegacyStorage` you can keep `PDO::*` convention.